### PR TITLE
CustomVocabLoader tests

### LINKED ISSUE #54: Add tests for VocabularyLoader
- [ ] move sample tables to `test/` folder
- [ ] cover edge cases mentioned [here](https://github.com/thehyve/ohdsi-etl-template/issues/41)
- [ ] test reading from single file with multiple vocabs vs reading multiple vocab-specific files

### PR DESCRIPTION
**Pull request by alepev**
Fixes #54

After adding tests, I refactored `CustomVocabLoader` to use a delegation pattern, which makes clearer what `<Table>Manager` instances are being used at each step in `load()`.

Still TODO:
- [x] minor docs update depending on how we decide to handle file prefixes for vocabs and classes

### CODE REVIEW

**Comment by alepev** [comment]
Tests complete, now starting refactoring to use delegation pattern in `CustomVocabLoader` instead of inheritance

File: tests/python/model/custom_vocab/test_custom_vocab_loader.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+        # (e.g. updated VOCAB4 in file with prefix of unchanged VOCAB3)
+
+        assert "VOCAB1_vocabulary.tsv contains vocabulary_ids that do not match file prefix:" \
+               " {'VOCAB2'}" in caplog.text
+        assert "VOCAB3_vocabulary.tsv contains vocabulary_ids that do not match file prefix:" \
+               " {'VOCAB4'}" in caplog.text
+        assert loaded_vocabs == ['VOCAB1_v2', 'VOCAB2_v2', 'VOCAB3_v1', 'VOCAB4_v2']
+
+        # classes are updated irrespective of the file prefix
+        assert loaded_classes == ['CLASS1_v2', 'CLASS2_v2']
```
Comment (alepev) [inline tests/python/model/custom_vocab/test_custom_vocab_loader.py:291]: So we agree with this behavior? At the moment file prefixes are kind of ignored when loading custom vocabularies and classes (though for vocabularies, you do get a warning of the vocabulary_id is different than the prefix). We might want to make it more strict and completely skip the file is the prefix does not match a vocabulary_id to be updated.

File: tests/python/model/custom_vocab/test_custom_vocab_loader.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        # prefix matching an updated vocabulary (VOCAB1) are processed
+        # (VOCAB3 is skipped). if file is processed, concept is loaded
+        # even if vocab not matching file prefix - but you get warnings
+        assert "VOCAB1_concept.tsv contains vocabulary_ids that do not match file prefix:" \
+               " {'VOCAB2'}" in caplog.text
+        # this file is skipped, hence no warning
+        assert "VOCAB3_concept.tsv contains vocabulary_ids that do not match file prefix:" \
+               " {'VOCAB4'}" not in caplog.text
+        assert loaded_vocabs == ['VOCAB1_v2', 'VOCAB2_v2', 'VOCAB3_v1', 'VOCAB4_v2']
+        assert loaded_concepts == [2000000001, 2000000002, 2000000005]
```
Comment (alepev) [inline tests/python/model/custom_vocab/test_custom_vocab_loader.py:317]: ...and here is the stricter concept behavior

**Review by Spayralbe — CHANGES_REQUESTED** [review]
The missing `continue` issue should be fixed (maybe also validate in a test that a <2B concept will not be committed), but otherwise great additions!

File: src/delphyne/model/custom_vocab/base_manager/concept_manager.py [hunk 3]
```diff
@@ -72,15 +98,27 @@ def _load_custom_concepts(self, vocab_ids: Set[str], valid_prefixes: Set[str]) -
...
-                            f'concept {concept_id} has duplicates across one or multiple '
-                            f'files')
+                        errors.add(f'concept {concept_id} is duplicated across one or multiple '
+                                   f'files')
+                        file_errors = True
+                        continue
+
+                    unique_concepts_check.add(concept_id)
 
                     if vocabulary_id in vocab_ids:
```
Comment (Spayralbe) [inline src/delphyne/model/custom_vocab/base_manager/concept_manager.py:123]: Could be useful to also log a warning when an unknown vocabulary_id (not present in any vocabulary.tsv file) is encountered in a concept file. Now they are silently ignored.

File: tests/python/model/custom_vocab/test_custom_vocab_loader.py (hunk 1 above)
Comment (Spayralbe) [reply tests/python/model/custom_vocab/test_custom_vocab_loader.py:289]: I think it's ok like this. Relation-wise the concept classes are not specific to a vocabulary anyway.

File: src/delphyne/model/custom_vocab/base_manager/class_manager.py [hunk 4]
```diff
@@ -177,29 +198,32 @@ def _load_custom_classes(self) -> None:
...
-                    elif class_id not in self._custom_classes_to_update:
-                        ignored_classes.update([class_id])
+                    elif class_id not in self.custom_classes_to_update:
+                        ignored_classes.add(class_id)
 
         if ignored_classes:
             logger.info(f'Skipped records with concept_class_id values that '
                         f'were already loaded under the current name: '
-                        f'{ignored_classes.most_common()}')
+                        f'{ignored_classes}')
```
[the remark this answers is not in the document]
Comment (alepev) [reply src/delphyne/model/custom_vocab/base_manager/class_manager.py]: agreed, updated here and in vocab_manager. updated tests to use debug-level caplog.

File: src/delphyne/model/custom_vocab/base_manager/concept_manager.py (hunk 3 above)
Comment (alepev) [reply src/delphyne/model/custom_vocab/base_manager/concept_manager.py:132]: Added warning and updated tests

File: src/delphyne/model/custom_vocab/base_manager/concept_manager.py [hunk 5]
```diff
@@ -72,15 +98,27 @@ def _load_custom_concepts(self, vocab_ids: Set[str], valid_prefixes: Set[str]) -
...
+                    # NOTE: only performing checks not already enforced
+                    # by CONCEPT table constraints; those will raise an
+                    # exception as soon as they occur
                     if int(concept_id) < 2000000000:
-                        raise ValueError(
-                            f'{concept_file.name} must have concept_ids starting at '
-                            f'2\'000\'000\'000 (2B+ convention)')
+                        errors.add(f'{concept_file.name} must contain concept_ids starting at '
+                                   f'2\'000\'000\'000 (2B+ convention)')
+                        file_errors = True
```
[the remark this answers is not in the document]
Comment (alepev) [reply src/delphyne/model/custom_vocab/base_manager/concept_manager.py:122]: That was a big omission, thanks for spotting!

File: src/delphyne/model/custom_vocab/base_manager/concept_manager.py [hunk 6]
```diff
@@ -11,7 +11,7 @@
 logger = logging.getLogger(__name__)
 
 
-class BaseConceptManager:
+class ConceptManager:
```
[the remark this answers is not in the document]
Comment (alepev) [reply src/delphyne/model/custom_vocab/base_manager/concept_manager.py:14]: Referenced in new issue. Agreed that is not ideal, but will pass on it for now (low priority).

**Comment by alepev** [comment]
OK I'm done now, I had forgotten to update the docs with how we process file prefixes

**Review by Spayralbe — APPROVED** [review]
🎍

**Comment by alepev** [comment]
🤩
